File: 0501-1000/0815. Bus Routes.java
// Hard
// BFS
// Time:O(routes+source),Space:O(routes+source)
// https://leetcode.cn/problems/bus-routes/

import java.util.*;

class Solution {
    // BFS: shortest path problem -> use Queue to go through
    // Map store each station and its list of routes passing through
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if (source == target)
            return 0;
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < routes.length; i++) {
            for (int station : routes[i]) {
                map.putIfAbsent(station, new ArrayList<>());
                map.get(station).add(i);
            }
        }
        boolean[] visStations = new boolean[100000], visRoutes = new boolean[routes.length];
        Queue<Integer> stationQue = new LinkedList<>();
        Queue<Integer> routeQue = new LinkedList<>();

        stationQue.add(source);
        routeQue.add(0);
        visStations[source] = true;
        int level = 0;

        while (!stationQue.isEmpty()) {
            int size = stationQue.size();
            for (int i = 0; i < size; i++) {
                int station = stationQue.poll();
                // check all the routes passing through this station
                // ensure cur station has routes before iterating over them
                List<Integer> routesForStation = map.get(station);
                if (routesForStation != null) {
                    for (int route : routesForStation) {
                        if (!visRoutes[route]) {
                            visRoutes[route] = true;
                            // add all stations in this route to queue
                            for (int nextStation : routes[route]) {
                                if (!visStations[nextStation]) {
                                    if (nextStation == target) {
                                        return level + 1;
                                    }
                                    stationQue.offer(nextStation);
                                    visStations[nextStation] = true;
                                }
                            }
                        }
                    }
                }
            }
            level++;
        }
        return -1;
    }
}

File: 0501-1000/0815. Bus Routes.py
# Time:O(N*S), Space:O(N*S)

from collections import defaultdict, deque
from typing import List

class Solution:
    def numBusesToDestination(
        self, routes: List[List[int]], source: int, target: int
    ) -> int:
        if source == target:
            return 0

        stop_to_routes = defaultdict(list)
        for i, route in enumerate(routes):
            for stop in route:
                stop_to_routes[stop].append(i)

        que = deque()
        vis = set()

        for route_id in stop_to_routes[source]:
            que.append((route_id, 1))
            vis.add(route_id)

        while que:
            route_id, buses = que.popleft()

            if target in routes[route_id]:
                return buses

            for stop in routes[route_id]:
                for next_route in stop_to_routes[stop]:
                    if next_route not in vis:
                        vis.add(next_route)
                        que.append((next_route, buses + 1))

        return -1
    